### PR TITLE
Fix replicating template messages

### LINKED ISSUE #787: Footer message that keeps replicating
Even if a footer message has already been added to the file specified in config.base, if I execute the script to add a change log for the newly developed part, a footer message will be added one by one every time.

- before
  ```changelog.md
  This Changelog was automatically generated by...
  ```
- after
  ```changelog.md
  This Changelog was automatically generated by...
  This Changelog was automatically generated by...
  ```
- after 2 times (keep replicating infinitely)
  ```changelog.md
  This Changelog was automatically generated by...
  This Changelog was automatically generated by...
  This Changelog was automatically generated by...
  ```

Maybe, this code doesn't work:
https://github.com/github-changelog-generator/github-changelog-generator/blob/af4e869d7d55326fe1c81842b7a2bbe4fc71be1e/lib/github_changelog_generator/generator/generator.rb#L62

- Environment
  - macOS Catalina (10.15.4)
  - project: [giturl](https://github.com/shinyaohtani/giturl/tree/master/)
  - ruby: ruby 2.7.1p83 (2020-03-31 revision a0c7c23c9c) [x86_64-darwin19]

### PR DESCRIPTION
**Pull request by shinyaohtani**
Fixes #787

### Header, footer and '\n' will be treated correctly
- Stop generating headers and footers indefinitely, leaving only one right each time.
  - To find the exact same string as the key, the Regex key must be escaped with Regex.escape().
  - **Initially I was trying to fix only credit_line. However, I have found that this framework needs to be applied not only to credit_line, but also to headers and front matter, so I modified them with same algorithm correctly.**
- Automatically remove only the automatically added '\n' and not remove any additional '\n' that may have been added manually by the user.
- specs for this issue is prepared.

### CODE REVIEW

**Comment by olleolleolle** [comment]
Thanks for this! I'll review soon. :+1:

File: lib/github_changelog_generator/generator/generator.rb [hunk 1]
```diff
@@ -151,5 +148,25 @@ def fetch_issues_and_pr
       add_first_occurring_tag_to_prs(@sorted_tags, @pull_requests)
       nil
     end
+
+    # Remove the previously assigned fixed message.
+    # @param [String] log Old lines are mixed
```
Comment (olleolleolle) [inline lib/github_changelog_generator/generator/generator.rb:153]:
```suggestion
    # @param log [String] Old lines are fixed
```

- a typo 🙂 
- newer YARD style: naming the parameter before the type description

File: lib/github_changelog_generator/generator/generator.rb [hunk 2]
```diff
@@ -151,5 +148,25 @@ def fetch_issues_and_pr
...
+    def remove_old_fixed_string(log)
+      log.gsub!(/#{Regexp.escape(@options[:frontmatter])}/, "") if @options[:frontmatter]
+      log.gsub!(/#{Regexp.escape(@options[:header])}\n{,2}/, "")
+      log.gsub!(/\n{,2}#{Regexp.escape(CREDIT_LINE)}/, "") # Remove old credit lines
+      log
+    end
+
+    # Add template messages
+    # @param [String] log Old lines are removed.
```
Comment (olleolleolle) [inline lib/github_changelog_generator/generator/generator.rb:162]:
```suggestion
    # Add template messages to given string. Previously added
    # messages of the same wording are removed.
    # @param log [String]
```

File: spec/unit/generator/generator_spec.rb [hunk 3]
```diff
@@ -0,0 +1,46 @@
+# frozen_string_literal: true
+
+require "spec_helper"
+require "./lib/github_changelog_generator/generator/generator"
```
Comment (olleolleolle) [inline spec/unit/generator/generator_spec.rb:4]:
```suggestion
require "github_changelog_generator/generator/generator"
```

File: spec/unit/generator/generator_spec.rb [hunk 4]
```diff
@@ -0,0 +1,46 @@
...
+      end
+    end
+  end
+
+  context "when plain contents string was given" do
+    describe "#insert_fixed_string" do
+      it "append template messages at header and footer" do
+        log = String.new(content)
+        ans = "#{header}\n\n#{content}\n\n#{footer}"
+        expect(generator.send(:insert_fixed_string, log)).to eq ans
```
Comment (olleolleolle) [inline spec/unit/generator/generator_spec.rb:42]:
```suggestion

        expect(generator.send(:insert_fixed_string, log)).to eq ans
```

File: spec/unit/generator/generator_spec.rb [hunk 5]
```diff
@@ -0,0 +1,46 @@
...
+      by [github_changelog_generator]\
+      (https://github.com/github-changelog-generator/github-changelog-generator)*
+    CREDIT
+  end
+
+  context "when the given base file has previously appended template messages" do
+    describe "#remove_old_fixed_string" do
+      it "removes old template headers and footers" do
+        log = +"#{header}\n\n#{header}\n#{header}#{content}\n\n#{footer}\n#{footer}#{footer}"
+        expect(generator.send(:remove_old_fixed_string, log)).to eq content
```
Comment (olleolleolle) [inline spec/unit/generator/generator_spec.rb:32]:
```suggestion

        expect(generator.send(:remove_old_fixed_string, log)).to eq content
```

**Comment by shinyaohtani** [comment]
I pressed all "Commit suggestion"s by @olleolleolle.

**Comment by shinyaohtani** [comment]
@olleolleolle 
Could release new version including this?

**Comment by shinyaohtani** [comment]
@olleolleolle 
May I release by myself following this document?
https://github.com/github-changelog-generator/github-changelog-generator/blob/master/CONTRIBUTING.md#releasing-a-new-version

**Pull request merged**